File: src/ruby/tormenta20/models/armadura.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing an Armadura (Armor) in Tormenta20.
    #
    # Armors provide defense bonuses but may impose armor penalties.
    # They are categorized as light (leve) or heavy (pesada).
    #
    # @example Find all light armors
    #   Tormenta20.armaduras.leves
    #
    # @example Find heavy armors sorted by defense
    #   Tormenta20.armaduras.pesadas.order(defense_bonus: :desc)
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g., "couro")
    # @!attribute [rw] name
    #   @return [String] Armor name in Portuguese
    # @!attribute [rw] category
    #   @return [String] Armor category: "leve" or "pesada"
    # @!attribute [rw] price
    #   @return [String] Price in Tormenta currency
    # @!attribute [rw] defense_bonus
    #   @return [Integer] Defense bonus provided by the armor
    # @!attribute [rw] armor_penalty
    #   @return [Integer] Armor penalty for skill checks
    # @!attribute [rw] weight
    #   @return [String] Weight in slots/spaces
    # @!attribute [rw] properties
    #   @return [Array<String>] Special armor properties
    # @!attribute [rw] description
    #   @return [String, nil] Optional description
    class Armadura < Base
      self.table_name = "armaduras"

      # @return [Array<String>] Valid armor categories
      CATEGORIES = %w[leve pesada].freeze

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true
      validates :category, presence: true, inclusion: { in: CATEGORIES }
      validates :defense_bonus, presence: true

      # @!method leves
      #   Filter light armors only.
      #   @return [ActiveRecord::Relation<Armadura>]
      scope :leves, -> { where(category: "leve") }

      # @!method pesadas
      #   Filter heavy armors only.
      #   @return [ActiveRecord::Relation<Armadura>]
      scope :pesadas, -> { where(category: "pesada") }

      # @!method by_category(category)
      #   Filter armors by category.
      #   @param category [String] Category: "leve" or "pesada"
      #   @return [ActiveRecord::Relation<Armadura>]
      scope :by_category, ->(c) { where(category: c) }

      # Check if this is a light armor.
      #
      # @return [Boolean] true if armor category is "leve"
      def leve?
        category == "leve"
      end

      # Check if this is a heavy armor.
      #
      # @return [Boolean] true if armor category is "pesada"
      def pesada?
        category == "pesada"
      end

      # Convert the armor to a hash representation.
      #
      # @return [Hash] Hash with all armor attributes (nil values excluded)
      def to_h
        {
          id: id,
          name: name,
          category: category,
          price: price,
          defense_bonus: defense_bonus,
          armor_penalty: armor_penalty,
          weight: weight,
          properties: properties,
          description: description
        }.compact
      end
    end
  end
end

File: src/ts/models/armadura.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"
import { getBookReference, type BookReference } from "./concerns/book_referenceable.js"

const TABLE = "armaduras"

/** Weight class of an armor. */
export type ArmaduraCategory = "leve" | "pesada"

/**
 * An Armadura (armor) in Tormenta20.
 *
 * @example
 * ```ts
 * Armadura.leves().all()
 * Armadura.pesadas().all()
 * Armadura.find("cota_de_malha")
 * ```
 */
export class Armadura extends BaseModel {
  get category(): ArmaduraCategory { return this._row.category as ArmaduraCategory }
  /** Price in tibares. */
  get price(): number { return (this._row.price as number) ?? 0 }
  /** Bonus added to Defense. */
  get defenseBonus(): number { return this._row.defense_bonus as number }
  /** Penalty applied to Agility-based skills and attack rolls. */
  get armorPenalty(): number { return (this._row.armor_penalty as number) ?? 0 }
  /** Weight in spaces (espaços). */
  get weight(): number { return (this._row.weight as number) ?? 0 }
  /** JSON array of special property objects. */
  get properties(): unknown[] { return BaseModel.json(this._row.properties, []) }
  get description(): string | null { return (this._row.description as string) ?? null }

  get isLeve(): boolean { return this.category === "leve" }
  get isPesada(): boolean { return this.category === "pesada" }

  /** Returns the sourcebook page reference, or `null` if not indexed. */
  bookReference(): BookReference | null { return getBookReference(this.id) }

  toH() {
    return {
      id: this.id, name: this.name, category: this.category, price: this.price,
      defense_bonus: this.defenseBonus, armor_penalty: this.armorPenalty,
      weight: this.weight, properties: this.properties, description: this.description,
    }
  }

  static query(): Query<Armadura> { return BaseModel.makeQuery(TABLE, Armadura) }
  static all(): Armadura[] { return Armadura.query().all() }
  static find(id: string): Armadura | null { return Armadura.query().find(id) }
  static first(): Armadura | null { return Armadura.query().first() }
  static last(): Armadura | null { return Armadura.query().last() }
  static count(): number { return Armadura.query().count() }

  static byCategory(cat: ArmaduraCategory): Query<Armadura> { return Armadura.query().where("category = ?", cat) }
  static leves(): Query<Armadura> { return Armadura.byCategory("leve") }
  static pesadas(): Query<Armadura> { return Armadura.byCategory("pesada") }
}
